### PR TITLE
Custom Status E2E tests

### PR DESCRIPTION
**Pull request by chetanyakan**
#### Summary

- Added cypress tests for status dropdown and custom status modal
- Added two support functions in cypress with documentation
- Added e2e tests for Custom Status modal and status dropdown according to spec.
- Added documentation above the custom status modal file and moved it to match the group
- Added CSS classes in custom status modal

#### Related Links

- Setting a default custom status, setting own custom status, clearing statuses and recent statuses:
https://automation-test-cases.vercel.app/test/MM-T3836
https://automation-test-cases.vercel.app/test/MM-T3846
https://automation-test-cases.vercel.app/test/MM-T3847

- Verifies the status appears in the right places when set, a status access point for new users:
https://automation-test-cases.vercel.app/test/MM-T3850
https://automation-test-cases.vercel.app/test/MM-T3851
https://automation-test-cases.vercel.app/test/MM-T3852

#### Related Pull Requests

- Has redux changes: https://github.com/mattermost/mattermost-redux/pull/1369

#### Screenshots

1. `cypress/integration/custom_status/custom_status_1_spec.js`
image

2. `cypress/integration/custom_status/custom_status_2_spec.js`
image

3. `cypress/integration/custom_status/custom_status_3_spec.js`
image

4. `cypress/integration/custom_status/custom_status_4_spec.js`

5. `cypress/integration/custom_status/custom_status_5_spec.js`
image

6. `cypress/integration/custom_status/custom_status_6_spec.js`
image

### CODE REVIEW

**Review by devinbinnie — CHANGES_REQUESTED** [review]
Thanks for the PR @chetanyakan!
I've added some comments for you to review.

File: e2e/cypress/integration/menus/status_dropdown_spec.js
Comment (hmhealey) [inline e2e/cypress/integration/menus/status_dropdown_spec.js:204]: What is `specify` used for? Is it the same as `it`?

File: e2e/cypress/integration/menus/status_dropdown_spec.js
Comment (hmhealey) [inline e2e/cypress/integration/menus/status_dropdown_spec.js:211]: Could you add a comment to this one?

Also, do we need to be this specific with the selector? Would something like `.status-dropdown-menu #custom_status__clear` work here instead, or do you think it's better to be more specific here? I could go either way with that.

File: components/custom_status/custom_status_modal.tsx
Comment (chetanyakan) [reply components/custom_status/custom_status_modal.tsx]: These classes were originally added to easily test that the custom statuses are moved from suggestions to recent statuses on satisfying appropriate conditions (i.e. not related to a different issue). 

This is now changed to use `id` instead of `className`.

File: e2e/cypress/integration/menus/status_dropdown_spec.js
Comment (chetanyakan) [reply e2e/cypress/integration/menus/status_dropdown_spec.js]: Yes. Both `specify` and `it` have the same meaning.

The reason `specify` was used here instead of `it` was because starting the sentence with `clear button should...` with `it` was not meaningful (i.e. `it clear button should...`).

**Comment by chetanyakan** [comment]
Rebased to HEAD of master.

File: package-lock.json
Comment (devinbinnie) [inline package-lock.json:15340]: Still need to revert these changes.

**Review by devinbinnie — CHANGES_REQUESTED** [review]
Thanks again @chetanyakan, this is looking good.
Just one more comment before I approve.

File: package-lock.json
Comment (chetanyakan) [reply package-lock.json]: Done.

**Review by devinbinnie — APPROVED** [review]
Looks good now, thanks @chetanyakan!

File: e2e/cypress/integration/menus/status_dropdown_spec.js
Comment (hmhealey) [reply e2e/cypress/integration/menus/status_dropdown_spec.js]: Oh I see. I didn't realize that was the reason some testing frameworks used `it` to begin with. That's actually kind of nice if we were consistent with it. Thanks for changing it to `it` though to keep that.

**Review by hmhealey — APPROVED** [review]

File: e2e/cypress/integration/custom_status/custom_status_spec.js
Comment (josephbaylon) [inline e2e/cypress/integration/custom_status/custom_status_spec.js:10]:
```suggestion
```

File: e2e/cypress/integration/custom_status/custom_status_spec.js
Comment (josephbaylon) [inline e2e/cypress/integration/custom_status/custom_status_spec.js:27]: Please move to its own file `custom_status_1_spec.js`

```suggestion
    describe('Custom Status - CTAs for New Users', () => {
```

File: e2e/cypress/integration/custom_status/custom_status_spec.js
Comment (josephbaylon) [inline e2e/cypress/integration/custom_status/custom_status_spec.js:61]: Please move to its own file `custom_status_2_spec.js`

```suggestion
    describe('Custom Status - Setting a Custom Status', () => {
```

File: e2e/cypress/integration/custom_status/custom_status_spec.js
Comment (josephbaylon) [inline e2e/cypress/integration/custom_status/custom_status_spec.js:175]: Please move to its own file `custom_status_3_spec.js`

```suggestion
    describe('Custom Status - Setting Your Own Custom Status', () => {
```

File: e2e/cypress/integration/custom_status/custom_status_spec.js
Comment (josephbaylon) [inline e2e/cypress/integration/custom_status/custom_status_spec.js:294]: Please move to its own file `custom_status_4_spec.js`

```suggestion
    describe('Custom Status - Recent Statuses', () => {
```

File: e2e/cypress/integration/custom_status/custom_status_spec.js
Comment (josephbaylon) [inline e2e/cypress/integration/custom_status/custom_status_spec.js:389]: Please move to its own file `custom_status_5_spec.js`

```suggestion
    describe('Custom Status - Verifying Where Custom Status Appears', () => {
```

File: e2e/cypress/integration/custom_status/custom_status_spec.js
Comment (josephbaylon) [inline e2e/cypress/integration/custom_status/custom_status_spec.js:521]: Please move to its own file `custom_status_6_spec.js`

```suggestion
    describe('Custom Status - Slash Commands', () => {
```

**Review by josephbaylon — CHANGES_REQUESTED** [review]
Hi @chetanyakan ! Thanks for your contribution !! Here are some initial change requests.

The test file is huge and we want to try to minimize the number of tests in a file. Please break up the tests into several files as suggested below and don't include the `// Stage: @prod` into the new files.

Also, I'm getting several failures locally. Not sure which is causing which. Please make sure to run each separated test locally, grab a snapshot of each successful run and please include the snapshots in the PR description.

Thanks! I'll re-review once the above changes are committed.

**Comment by chetanyakan** [comment]
@josephbaylon Test case 4 has a weird behaviour.

It runs fine from the terminal, but fails when running from the cypress desktop app.


image


image

**Comment by chetanyakan** [comment]
Strangely, I get different results using different version of Cypress.

@josephbaylon Can you verify on your end if any of the custom status E2E tests are failing?

**Review by josephbaylon — APPROVED** [review]
Hi @chetanyakan I retested locally and all tests pass for me locally! Thanks.

However, given that this is behind a feature flag, we cannot merge this until we set the value to true by default.

@ogi-m I'll approve the PR but I'll set the `DO NOT MERGE` label for now. We can revisit once the feature flag is enabled by default.

**Review by ogi-m — APPROVED** [review]
@josephbaylon The feature flag is now enabled on master by default, can it be merged now?

**Comment by josephbaylon** [comment]
@hmhealey It seems like I lost rights to sync and merge. Please sync and merge. Thanks!

**Comment by saturninoabril** [comment]
@chetanyakan Reviews complete. Please update so we can merge. Thank you for your contribution!

**Comment by chetanyakan** [comment]
@saturninoabril @josephbaylon Updated the branch.

**Pull request merged**
